### PR TITLE
protobuf: remove most getters

### PR DESCRIPTION
**Pull request by tamird**
Contains some cleanup missed by #2875, /cc @thschroeter.

I'd review this commit by commit - the 2nd commit does the work for all but `roachpb.Value`, which is then done in the 3rd commit.

@tschottdorf you're the lucky winner! Sorry. @bdarnell I'd appreciate your eyes on this as well.

### CODE REVIEW

**Comment by bdarnell** [comment]
What's the motivation for removing all getters?

**Comment by tschottdorf** [comment]
LGTM, though I skipped lightly over the tests and must've missed that we don't like getters.

**Comment by tamird** [comment]
The motivation for removing getters was that they got in my way when I was looking into usage of `.RawBytes`. It then turned out that we don't really use them very much, and that the nil protection they provide is generally not needed or wanted (since it can paper over bugs).

**Comment by bdarnell** [comment]
By "got in your way" you mean you needed to grep for both `.RawBytes` and `GetRawBytes`? 

Does this change fix any bugs that were being "papered over", or is that a hypothetical concern? 

The getters help avoid data-dependent panics, which seems a worthy goal. For nullable fields, I think we should encourage the use of the getters where applicable rather than ban them. (For non-nullable fields the getters are redundant and we might as well save a little code size by getting rid of them).

**Comment by tamird** [comment]
> The getters help avoid data-dependent panics, which seems a worthy goal. For nullable fields, I think we should encourage the use of the getters where applicable rather than ban them. (For non-nullable fields the getters are redundant and we might as well save a little code size by getting rid of them).

The way in which these getters help avoid panics is by implicitly assuming that the absence of a value is equivalent to the presence of the zero value. This might be reasonable to assume in the context of golang/protobuf, but it is not reasonable to assume in the context of gogo/protobuf which allows us to customize our structures for the desired semantics.

The only panics prevented by getters that I noticed while writing this PR were prevented by the getters' receiver-is-nil protection.

I'd encourage you to look at the actual changes that resulted here; I don't believe I made any substantial function signature changes (as would be necessary in the case of a value-returning getter being replaced by nullable field access).

**Comment by bdarnell** [comment]
> The only panics prevented by getters that I noticed while writing this PR were prevented by the getters' receiver-is-nil protection.

The receiver-is-nil protection is exactly the kind of data-dependent panic I was referring to. The getters can be used to chain accesses through a series of fields that may or may not be present.

**Comment by tschottdorf** [comment]
I agree in principle, but we've been using getters when we weren't really sure whether something could be `nil`. I went through all the non-testing code in the diff and found that there were only a handful places where that was really possible, and in all others we didn't need to worry in the first place.

If something uses getters heavily, we can always re-add them (for example, Tamir kept them for the commit triggers).

**Comment by tamird** [comment]
There are 6 instances of `().` in the diff (indicative of chaining). They break down as follows:
- 3 removals of getter calls for non-nullable fields
- 3 replacements with nil checks (2 of these are commit trigger related)

I didn't look for similar cases without chaining (i.e. someone using a temporary which might be nil), but I think this is still a strong case for adding getters only when we really want them.

**Comment by bdarnell** [comment]
I admit that the case for adding getters if they did not already exist would be weak (given the existence of non-nullable fields). But the argument for removing them is also weak. If a change like this requires scrutiny of individual call sites and is not fixing a demonstrated problem, it's not worth making IMO.

**Comment by tschottdorf** [comment]
Agreed that it's somewhat borderline, but I just want to mention that usage of a `Getter` in the code automatically leads me to believe that the `Gettee` can actually be `nil`. More often than not, it's not the case, and the author just didn't want to have to think about it.
So, from my personal experience, I'd rather not have `Getters` anywhere unless they're vouched for. That's pretty much this change in a nutshell.

**Comment by tamird** [comment]
> But the argument for removing them is also weak. If a change like this requires scrutiny of individual call sites and is not fixing a demonstrated problem, it's not worth making IMO.

I don't think this change requires scrutiny for the purpose of correctness; my previous post was scrutinizing the individual call sites to demonstrate that the ergonomics were not being adversely affected.

**Comment by bdarnell** [comment]
OK. I still think that more time has been spent making and reviewing this change than will ever be saved by its inclusion, but y'all care about this more than I do so go ahead.

**Comment by tamird** [comment]
Hey, at least this resulted in our first "y'all" https://github.com/cockroachdb/cockroach/issues?q=y%27all

File: config/config.go [hunk 1]
```diff
@@ -104,17 +104,16 @@ func ObjectIDForKey(key roachpb.RKey) (uint32, bool) {
 
 // GetValue searches the kv list for 'key' and returns its
 // raw byte value if found. ok is true only if the key is found.
```
Comment (petermattis) [inline config/config.go]: This comment needs updating.

File: storage/replica_raftstorage.go [hunk 2]
```diff
@@ -193,8 +192,13 @@ func (r *Replica) loadAppliedIndex(eng engine.Engine) (uint64, error) {
...
 	}
 	if v != nil {
-		var err error
-		_, appliedIndex, err = encoding.DecodeUint64(v.GetRawBytes())
+		bytes, err := v.GetBytes()
+		if err != nil {
+			return 0, err
+		}
+		// Cannot use v.GetInt() here because we need the precision of
+		// uint64; GetInt() is lossy.
```
Comment (petermattis) [inline storage/replica_raftstorage.go]: How is `GetInt()` lossy? It decodes using `encoding.DecodeUint64()` and casts to an `int64`. I think you should be able to cast it back to a `uint64` if desired.

Comment (petermattis) [inline storage/replica_raftstorage.go]: Related: I think `setAppliedIndex` should use `roachpb.Value.SetInt(int64(appliedIndex))`.

Comment (tamird) [inline storage/replica_raftstorage.go]: > It decodes using `encoding.DecodeUint64()` and casts to an `int64`.

What if the decoded value overflows `int64`?

Comment (petermattis) [inline storage/replica_raftstorage.go]: http://play.golang.org/p/mkf-UdFuRp

File: config/config.go (hunk 1 above)
Comment (tamird) [inline config/config.go]: https://github.com/cockroachdb/cockroach/pull/2917

File: storage/replica_raftstorage.go (hunk 2 above)
Comment (tamird) [inline storage/replica_raftstorage.go]: https://github.com/cockroachdb/cockroach/pull/2917

**Pull request merged**